subworkflows/local/utils_nfcore_raredisease_pipeline/main.nf

```diff
@@ -0,0 +1,279 @@
+    //
+    workflow.onComplete {
+        if (email || email_on_fail) {
+            completionEmail(summary_params, email, email_on_fail, plaintext_email, outdir, monochrome_logs, multiqc_report.toList())
+        }
+
+        completionSummary(monochrome_logs)
+
+        if (hook_url) {
+            imNotification(summary_params, hook_url)
+        }
+    }
+}
+
+/*
+========================================================================================
+    FUNCTIONS
+========================================================================================
+*/
+//
+// Check and validate pipeline parameters
+//
+def validateInputParameters() {
+    genomeExistsError()
+}
+
+//
+// Validate channels from input samplesheet
+//
+def validateInputSamplesheet(input) {
+    def (metas, fastqs) = input[1..2]
+
+    // Check that multiple runs of the same sample are of the same datatype i.e. single-end / paired-end
+    def endedness_ok = metas.collect{ it.single_end }.unique().size == 1
+    if (!endedness_ok) {
+        error("Please check input samplesheet -> Multiple runs of a sample must be of the same datatype i.e. single-end or paired-end: ${metas[0].id}")
+    }
+
+    return [ metas[0], fastqs ]
+}
+//
+// Get attribute from genome config file e.g. fasta
+//
+def getGenomeAttribute(attribute) {
+    if (params.genomes && params.genome && params.genomes.containsKey(params.genome)) {
+        if (params.genomes[ params.genome ].containsKey(attribute)) {
+            return params.genomes[ params.genome ][ attribute ]
+        }
+    }
+    return null
+}
+//
+// Replace spaces in vcf files with underscores
+//
+def replaceSpacesInInfoColumn(vcf_file, parent_dir, base_name) {
+    def outfile = new File(parent_dir + '/' + base_name + '_formatted.vcf')
+    def writer  = outfile.newWriter()
+    vcf_file.eachLine { line ->
+        if (line.startsWith("#")) {
+            writer << line + "\n"
+        } else {
+            def split_str = line.tokenize("\t")
+            split_str[7] = split_str.getAt(7).replaceAll(" ","_")
+            writer << split_str.join("\t") + "\n"
+        }
+    }
+    writer.close()
+    return outfile
+}
+
+//
+// Exit pipeline if incorrect --genome key provided
+//
+def genomeExistsError() {
+    if (params.genomes && params.genome && !params.genomes.containsKey(params.genome)) {
+        def error_string = "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
+            "  Genome '${params.genome}' not found in any config files provided to the pipeline.\n" +
+            "  Currently, the available genome keys are:\n" +
+            "  ${params.genomes.keySet().join(", ")}\n" +
+            "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+        error(error_string)
+    }
+}
+
+//
+// Generate methods description for MultiQC
+//
+def toolCitationText() {
+    // TODO nf-core: Optionally add in-text citation tools to this list.
```

Comment:
There are three TODOs left in `subworkflows/local/utils_nfcore_raredisease_pipeline/main.nf`. I don't know whether it is important.

Reply:
We have the list of citations in CITATIONS.md already, so I think its better to keep just that one updated as we go. I will remove the TODO statements 😄